test/lib/util/aws/cloudfront_signed_url_test.exs

```diff
@@ -0,0 +1,96 @@
+  @signature_with_query1   "TzC4VcV7T1RRoiE32iSIzPxkD9X4-hHKrBWgLgDUuCay~Sqt3E1fb~z4eShZUw6-wGZIC3rFaYUJ9SswQZpaGe6mRUCscqBJQ0uilpMsxq2bZl6HU5sGAagD6tD9~MAlWs8Q74BUVs8CbUmsX2buKHvRdnvnmkhjM4KocFaOsIBeHWxb3M0L~lYokjRaSHs3NUcACne6vo0~AB6fwwfkJ1bvxBD2KHDzbMHY2NrKau0zBkgRHcofr9tRNa7fPT-rwrdNVDcViFwnf1i3D2PSInWeVJuD9s9w7gU4Wigv5Y0OkYION-ZZkEswPUVEgttJSOKrP8DHgL6LQ1b~Vbgg9g__"
+  @signature_with_query2   "TTfoKwPF7xn-ZRilOXwuidh1tAtJFJm8BPs6dBc4zLNGlgV~V7XN1YvDy5xNgZInUf1CvpvRMW0NFUqxJqNZd3FOk6QpySVoHbWYQWYoEH2eWV5cftYtJi2GUaol~yDLEN36zuOHmDSfYpSPktYIePiagz9e3R3ZzNEQUCpIZTO8XfJiISCoHkREeJxA4slMQhhuofFwSCfr2yW2ZMVk75JxPtNmAde-sj4C09PORyNWTuWOoYYghb5DfF4yDlk5WeQKCwGmjhsQ7mM7UMU3xk-fOKWdZcyxUIjqxlQSQEVVwP4QlHdFo4y2uNhER8e0HvhQk-Pbpfj~JFePmeNoPA__"
+  @signature_with_query3   "BzTb7kOa1P2sp2lO~sO4j17b7UZpxgDXIvR4ET5FFa0rwu9Z5u9ubC09EGjHn9r-k~pxWu-KJyvr--f0MDNtUGWR5fgOb74iobt4GP9-NsS9WRu3JdhFZXHqAGkPF9TtSXrP8qMtcsBJ~F0SWeWevdhU19FMPH4hIe9ff0m2J29BIFZUe9~fyWh3daMQ9aC7Zf84G3V00FW~v21ggcRhDgkLdN6eUCm-cayzs6kZhr0V6PABjw0Ve9hJbAHC-B2PbTwfFVkiLn0B4dQtgwYC9NHhvWzil7xGBhJvAX7CPJJNImi~A1x8FdigTgMUXg4fwAapj0tk4FmkFu-NHI-AQg__"
+
+  defp assert_query_params(query_params, expected_signature) do
+    [
+      {"Expires"    , expires    },
+      {"Signature"  , signature  },
+      {"Key-Pair-Id", key_pair_id},
+    ] = query_params
+    assert expires     == Integer.to_string(@expires_in_seconds)
+    assert signature   == expected_signature
+    assert key_pair_id == @key_pair_id
+  end
+
+  defp assert_base_url_and_get_query_params(url) do
+    %URI{host: host, path: path, query: query, scheme: scheme} = URI.parse(url)
+    assert "#{scheme}://#{host}#{path}" == URI.encode(@resource_url)
+    URI.query_decoder(query) |> Enum.to_list()
+  end
+
+  setup do
+    lifetime = 60
+    :meck.expect(System, :system_time, fn :second -> @expires_in_seconds - lifetime end)
+    on_exit(fn -> :meck.unload() end)
+    %{lifetime: lifetime}
+  end
+
+  test "should generate a signed URL for URL without query", %{lifetime: lifetime} do
+    signed_url = CloudfrontSignedUrl.generate_signed_url(@resource_url, lifetime, @key_pair_id, @private_key)
+    query_params = assert_base_url_and_get_query_params(signed_url)
+    assert_query_params(query_params, @signature_without_query)
+  end
+
+  test "should generate a signed URL for URL with query", %{lifetime: lifetime} do
+    [{"?", @signature_with_query1}, {"?hello=world", @signature_with_query2}, {~s/?hello="日本"&/, @signature_with_query3}]
+    |> Enum.each(fn {query, signature} ->
+      resource_url = @resource_url <> query
+      signed_url = CloudfrontSignedUrl.generate_signed_url(resource_url, lifetime, @key_pair_id, @private_key)
+      assert String.starts_with?(signed_url, URI.encode(resource_url) <> "&Expires=")
+      query_params = assert_base_url_and_get_query_params(signed_url)
+      query_params |> Enum.reverse() |> Enum.take(3) |> Enum.reverse() |> assert_query_params(signature)
```

Comment:
Same as https://github.com/access-company/antikythera/pull/100#discussion_r337681291.

Reply:
I fixed in https://github.com/access-company/antikythera/pull/100/commits/ff031d728170f0e6818b931ea7200e6a0d815f86